File: src/main/java/de/espend/idea/vuejs/linemarker/ComponentUsageLineMarker.java
package de.espend.idea.vuejs.linemarker;

import com.intellij.codeInsight.daemon.LineMarkerInfo;
import com.intellij.codeInsight.daemon.LineMarkerProvider;
import com.intellij.codeInsight.navigation.NavigationGutterIconBuilder;
import com.intellij.codeInsight.navigation.impl.PsiTargetPresentationRenderer;
import com.intellij.openapi.util.NotNullLazyValue;
import com.intellij.openapi.vfs.VfsUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.util.indexing.FileBasedIndex;
import de.espend.idea.vuejs.VueJsIcons;
import de.espend.idea.vuejs.index.ComponentUsageIndex;
import de.espend.idea.vuejs.utils.VueJsUtil;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.vuejs.VuejsIcons;
import org.jetbrains.vuejs.lang.html.VueFile;

import javax.swing.*;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class ComponentUsageLineMarker implements LineMarkerProvider {
    @Override
    public @Nullable LineMarkerInfo<?> getLineMarkerInfo(@NotNull PsiElement psiElement) {
        return null;
    }

    @Override
    public void collectSlowLineMarkers(@NotNull List<? extends PsiElement> psiElements, @NotNull Collection<? super LineMarkerInfo<?>> lineMarkerInfos) {
        // we need project element; so get it from first item
        if (psiElements.isEmpty()) {
            return;
        }

        for (PsiElement psiElement : psiElements) {
            if (psiElement instanceof VueFile vueFile && vueFile.getName().endsWith(".vue")) {
                if (vueFile.getName().endsWith(".vue")) {
                    String filenameWithoutExtension = vueFile.getName().substring(0, vueFile.getName().length() - 4);
                    if (FileBasedIndex.getInstance().getContainingFiles(ComponentUsageIndex.KEY, filenameWithoutExtension, GlobalSearchScope.allScope(psiElement.getProject())).isEmpty()) {
                        return;
                    }

                    NavigationGutterIconBuilder<PsiElement> builder = NavigationGutterIconBuilder.create(VueJsIcons.VUE_JS_TOOLBOX)
                        .setTooltipText("Vue.js Toolbox: Navigate to Usages")
                        .setTargetRenderer(MyFileReferencePsiElementListCellRenderer::new)
                        .setTargets(NotNullLazyValue.lazy(() -> {
                            Collection<PsiElement> elements = new ArrayList<>();

                            for (VirtualFile value : FileBasedIndex.getInstance().getContainingFiles(ComponentUsageIndex.KEY, filenameWithoutExtension, GlobalSearchScope.allScope(psiElement.getProject()))) {
                                PsiFile file = PsiManager.getInstance(psiElement.getProject()).findFile(value);
                                if (file instanceof VueFile vueFile1) {
                                    for (Map.Entry<String, List<String>> entry : FileBasedIndex.getInstance().getFileData(ComponentUsageIndex.KEY, value, file.getProject()).entrySet()) {
                                        String refImport = entry.getValue().get(0);
                                        String importAlias = entry.getValue().get(1);

                                        VirtualFile relativeFile = VfsUtil.findRelativeFile(refImport, value);

                                        if (vueFile.getVirtualFile().equals(relativeFile)) {
                                            elements.addAll(VueJsUtil.getTemplateTags(vueFile1, importAlias, VueJsUtil.convertToKebabCase(importAlias)));
                                        }
                                    }
                                }
                            }

                            return elements;
                        }));

                    lineMarkerInfos.add(builder.createLineMarkerInfo(psiElement));
                }
            }
        }
    }

    private static class MyFileReferencePsiElementListCellRenderer extends PsiTargetPresentationRenderer<PsiElement> {
        @Nls
        @Nullable
        @Override
        public String getContainerText(@NotNull PsiElement element) {
            return element.getContainingFile().getName();
        }

        @Nullable
        @Override
        protected Icon getIcon(@NotNull PsiElement element) {
            return VuejsIcons.Vue;
        }
    }
}

File: src/main/kotlin/de/espend/idea/vuejs/linemarker/ComponentUsageLineMarker.kt
package de.espend.idea.vuejs.linemarker

import com.intellij.codeInsight.daemon.LineMarkerInfo
import com.intellij.codeInsight.daemon.LineMarkerProvider
import com.intellij.codeInsight.navigation.NavigationGutterIconBuilder
import com.intellij.codeInsight.navigation.impl.PsiTargetPresentationRenderer
import com.intellij.javascript.nodejs.packageJson.PackageJsonFileManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.NotNullLazyValue
import com.intellij.openapi.vfs.VfsUtil
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiManager
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.util.indexing.FileBasedIndex
import de.espend.idea.vuejs.VueJsIcons
import de.espend.idea.vuejs.index.ComponentUsageIndex
import de.espend.idea.vuejs.utils.VueJsUtil
import org.jetbrains.vuejs.VuejsIcons
import org.jetbrains.vuejs.lang.html.VueFile
import javax.swing.Icon

class ComponentUsageLineMarker : LineMarkerProvider {
    override fun getLineMarkerInfo(psiElement: PsiElement): LineMarkerInfo<*>? = null

    override fun collectSlowLineMarkers(psiElements: List<PsiElement>, lineMarkerInfos: MutableCollection<in LineMarkerInfo<*>>) {
        if (psiElements.isEmpty()) return

        for (psiElement in psiElements) {
            if (psiElement is VueFile && psiElement.name.endsWith(".vue")) {
                val filenameWithoutExtension = psiElement.name.substring(0, psiElement.name.length - 4)
                if (FileBasedIndex.getInstance().getContainingFiles(ComponentUsageIndex.KEY, filenameWithoutExtension, GlobalSearchScope.allScope(psiElement.project)).isEmpty()) {
                    return
                }

                val builder = NavigationGutterIconBuilder.create(VueJsIcons.VUE_JS_TOOLBOX)
                    .setTooltipText("Vue.js Toolbox: Navigate to Usages")
                    .setTargetRenderer { MyFileReferencePsiElementListCellRenderer() }
                    .setTargets(NotNullLazyValue.lazy {
                        val elements = mutableListOf<PsiElement>()

                        for (value in FileBasedIndex.getInstance().getContainingFiles(ComponentUsageIndex.KEY, filenameWithoutExtension, GlobalSearchScope.allScope(psiElement.project))) {
                            val file = PsiManager.getInstance(psiElement.project).findFile(value)
                            if (file is VueFile) {
                                for ((_, entryValue) in FileBasedIndex.getInstance().getFileData(ComponentUsageIndex.KEY, value, file.project)) {
                                    val refImport = entryValue[0]
                                    val importAlias = entryValue[1]

                                    if (refImport.startsWith(".")) {
                                        val relativeFile = VfsUtil.findRelativeFile(refImport, value)
                                        if (psiElement.virtualFile == relativeFile) {
                                            elements.addAll(VueJsUtil.getTemplateTags(file, importAlias, VueJsUtil.convertToKebabCase(importAlias)))
                                        }
                                    } else {
                                        var replace = refImport.replace("\\", "/")
                                        if (replace.startsWith("~/")) replace = replace.substring(2)
                                        if (isImportInScope(file.project, psiElement.virtualFile, replace)) {
                                            elements.addAll(VueJsUtil.getTemplateTags(file, importAlias, VueJsUtil.convertToKebabCase(importAlias)))
                                        }
                                    }
                                }
                            }
                        }

                        elements
                    })

                lineMarkerInfos.add(builder.createLineMarkerInfo(psiElement))
            }
        }
    }

    companion object {
        private fun isImportInScope(project: Project, fileOpen: VirtualFile, foreignImport: String): Boolean {
            for (validPackageJsonFile in PackageJsonFileManager.getInstance(project).validPackageJsonFiles) {
                VfsUtil.getRelativePath(fileOpen, validPackageJsonFile.parent, '/') ?: continue
                val relativeFile = VfsUtil.findRelativeFile(foreignImport, validPackageJsonFile)
                if (fileOpen == relativeFile) return true
            }
            return false
        }
    }

    private class MyFileReferencePsiElementListCellRenderer : PsiTargetPresentationRenderer<PsiElement>() {
        override fun getContainerText(element: PsiElement): String? = element.containingFile.name
        override fun getIcon(element: PsiElement): Icon? = VuejsIcons.Vue
    }
}
